docs/custom_promise_types/groups.py

```diff
@@ -0,0 +1,335 @@
+
+        # if attribute policy is present
+        if "policy" in attributes:
+            policy = attributes["policy"]
+
+            # check attribute policy type
+            if type(policy) is not str:
+                raise ValidationError("invalid type for attribute policy: expected string")
+
+            # check attribute policy value
+            if policy != "present" and policy != "absent":
+                raise ValidationError(f"invalid value '{policy}' for attribute policy: must be 'present' or 'absent'")
+
+            # check attributes gid and members are not used with policy absent
+            if policy == "absent":
+                if "gid" in attributes:
+                    self._log_warning(f"cannot assign gid to absent group '{promiser}'")
+                if "members" in attributes:
+                    self._log_warning(f"cannot assign members to absent group '{promiser}'")
+        
+        # if attribute gid is present
+        if "gid" in attributes:
+            gid = attributes["gid"]
+
+            # check attribute gid type
+            if type(gid) is not str:
+                raise ValidationError("invalid type for attribute gid: expected string")
+
+            # check attribute gid value
+            if type(gid) == str:
+                try: 
+                    int(gid)
+                except ValueError:
+                    raise ValidationError(f"invalid value '{gid}' for attribute gid: expected integer literal")
+
+        # if attribute members is present
+        if "members" in attributes:
+            # some black magic which removes back slashes parsing before parsing json
+            members = json.loads(json.loads('"' + attributes["members"] + '"'))
+            attributes["members"] = members
+            
+            # check attribute only not used with attributes include or exclude
+            if "only" in members and ("include" in members or "exclude" in members):
+                raise ValidationError("attribute only may not be used with attributes include or exclude")
+            
+            # check attributes of attibutes in members
+            for attr in members:
+                if (attr != "only" and attr != "include" and attr != "exclude"):
+                    raise ValidationError(f"invalid value '{attr}' in attribute members: must be only, exclude or include")
+            
+            # check values of attributes in members
+            for users in members.values():
+                for user in users:
+                    match = re.match(self._name_regex, user)
+                    if match == None:
+                        self.log_warning(f"username '{user}' not according to recomendations in system reference manual")
+                    if len(user) > self._name_maxlen:
+                        raise ValidationError(f"username too long: ({len(user)} > {self._name_maxlen})")
```

Comment:
I don't think this should care about the validity of the usernames, they are or aren't users existing in the system, that's all we this needs to care about, right?

Reply:
Good point!